typing/includecore.ml

```diff
@@ -427,6 +459,131 @@ let compare_records_with_representation ~loc env params1 params2 n
        (Record_regular|Record_inlined _|Record_extension _) ->
         assert false
 
+end
+
+
+module Variant_diffing = struct
+
+  let compare_constructor_arguments ~loc env params1 params2 arg1 arg2 =
+    match arg1, arg2 with
+    | Types.Cstr_tuple arg1, Types.Cstr_tuple arg2 ->
+        if List.length arg1 <> List.length arg2 then
+          Some (Arity : constructor_mismatch)
+        else begin
+        (* Ctype.equal must be called on all arguments at once, cf. PR#7378 *)
+        match Ctype.equal env true (params1 @ arg1) (params2 @ arg2) with
+        | exception Ctype.Equality trace -> Some (Type (env, trace))
+        | () -> None
+      end
+    | Types.Cstr_record l1, Types.Cstr_record l2 ->
+        Option.map
+          (fun rec_err -> Inline_record rec_err)
+          (Record_diffing.compare env ~loc params1 params2 l1 l2)
+    | Types.Cstr_record _, _ -> Some (Kind First : constructor_mismatch)
+    | _, Types.Cstr_record _ -> Some (Kind Second : constructor_mismatch)
+
+  let compare_constructors ~loc env params1 params2 res1 res2 args1 args2 =
+    match res1, res2 with
+    | Some r1, Some r2 ->
+        begin match Ctype.equal env true [r1] [r2] with
+        | exception Ctype.Equality trace -> Some (Type (env, trace))
+        | () -> compare_constructor_arguments ~loc env [r1] [r2] args1 args2
+        end
+    | Some _, None -> Some (Explicit_return_type First)
+    | None, Some _ -> Some (Explicit_return_type Second)
+    | None, None ->
+        compare_constructor_arguments ~loc env params1 params2 args1 args2
+
+  let equal ~loc env params1 params2
+      (cstrs1 : Types.constructor_declaration list)
+      (cstrs2 : Types.constructor_declaration list) =
+    List.length cstrs1 = List.length cstrs2 &&
+    List.for_all2 (fun (cd1:Types.constructor_declaration)
+                    (cd2:Types.constructor_declaration) ->
+        Ident.name cd1.cd_id = Ident.name cd2.cd_id
+        &&
+        begin
+          Builtin_attributes.check_alerts_inclusion
+            ~def:cd1.cd_loc
+            ~use:cd2.cd_loc
+            loc
+            cd1.cd_attributes cd2.cd_attributes
+            (Ident.name cd1.cd_id)
+          ;
+        match compare_constructors ~loc env params1 params2
+                cd1.cd_res cd2.cd_res cd1.cd_args cd2.cd_args with
+        | Some _ -> false
+        | None -> true
+      end) cstrs1 cstrs2
+
+  let update _ () = ()
+
+  let weight = function
+    | Diffing.Insert _ -> 10
+    | Diffing.Delete _ -> 10
+    | Diffing.Keep _ -> 0
+    | Diffing.Change (_,_,Diffing_with_keys.Name t) ->
+        if t.types_match then 10 else 15
+    | Diffing.Change _ -> 10
```

Comment:
This function is a duplicate from the other `weight` function. And `default_weight` appears unused. Something is wrong here.

Reply:
Indeed, this is the `default_weight` function for `Diffing_with_keys`, which is currently missing.